### ISSUE TITLE
Configuration File Check

### ISSUE THREAD

**Issue by aaron-pham** [jdholtz/auto-southwest-check-in#207]
### Description

Is there any interest in adding a configuration file checker for changes?

I know there's the `retrieval_interval`, but that only checks new flights on accounts and not any new reservations added. I think it would be nice to add the feature of either:

- Watching the file for changes and re-reading in the file if changes are detected
- Using the retrieve interval to re-read the configuration file

These solutions could be naive, I haven't had a good chance to dig into the code yet. Based off of my current usage, I would kind of like option 1 - I'm essentially hosting a discord server for friends and every time they get a new flight, it would be nice to just add to the config file and have it automatically do all the checks.

Option 2 makes sense to me, when reading the readme I got the impression that this would pick up new reservations too. But after it "failed" to do that, I looked into a bit more and noticed it should just be the accounts. This could just be a me issue though.

Let me know what you think of these suggestions.

**Comment by jdholtz** [comment]
I did discuss this [here](https://github.com/jdholtz/auto-southwest-check-in/discussions/79#discussioncomment-5730029). Let me know your thoughts after reading that.

**Comment by aaron-pham** [comment]
I probably need to actually dig into the code more, but off initial impression I'd think this is how I'd implement it. 

1. Create the main global config as a singleton
2. When file changes are detected, parse the file & create the configuration "fresh" (some method to create a temporary instance, since the class will now be a singleton)
3. Compare this fresh configuration with what's in the global configuration - if it exist there, update it with the new configuration
4. Add any configurations that don't exist yet in this singleton global configuration

Once we update the configurations, on the next loop I believe the sub-processes will pick up the changes. The only thing is it won't be instant due to the sleep, but I think it's a decent enough middleground. 

Thoughts?

**Comment by jdholtz** [comment]
> Once we update the configurations, on the next loop I believe the sub-processes will pick up the changes. 

Unfortunately not. Since multiprocessing is used in the script, when each process is created, all memory copied to that process, so the changes will not propagate to the child processes.

I do agree that the part you outlined could be a way to detect config changes. However, the complicated part is restarting all the processes (this could be done by just killing all of them and restarting the script, which is essentially what I will detail below using other software). This is the main reason (as I pointed out in the discussion I linked above) as to why I don't think this would be a great addition within the script itself (it makes it much more complicated).

Fortunately, there is a very simple way to do this using third-party software. Using [entr](https://github.com/eradman/entr), you can do `ls config.json | entr -r ./watch.sh ARGUMENTS_TO_SCRIPT` with `watch.sh`:
```bash
pkill -f "python southwest.py"
python southwest.py "$@"
```
This will kill the current script and restart the process automatically once you change the config file.

**Comment by jdholtz** [comment]
> when reading the readme I got the impression that this would pick up new reservations too. But after it "failed" to do that, I looked into a bit more and noticed it should just be the accounts. This could just be a me issue though

Could you clarify what you mean here? The script will automatically schedule new reservations when they are added to an account (provided you gave account credentials to the script). When providing just reservation information, this isn't possible because a reservation is not tied to other reservations for the user (without using the account, obviously).

**Comment by aaron-pham** [comment]
> Since multiprocessing is used in the script, when each process is created, all memory copied to that process, so the changes will not propagate to the child processes.

Gotcha, I'm not too familiar with multiprocessing so thought the changes would propagate down. It's probably not a good idea to use multiprocessing.Manager (or other shared memory methods), huh.

> Fortunately, there is a very simple way to do this using third-party software. Using [entr](https://github.com/eradman/entr), you can do ls config.json | entr -r ./watch.sh ARGUMENTS_TO_SCRIPT with watch.sh:

The only "issue" I have with with this is I have notifications set up to the discord server and don't want to constantly have spam about the check-in being scheduled if the notification came through once already. But do definitely want the price check to continue to come. Any thoughts on this? I have some hackier solutions in mind (something like keeping state in a file somewhere on notifications sent), but I doubt that would be healthy for this repo 😂.

> Could you clarify what you mean here? 

Sorry, I worded this badly. I was thinking that if I set the `retrieve interval`, it would be checking the configuration file for changes, rather than account level changes. This is more of a me reading the documentation quickly issue than a real issue. I just mentioned it as more of a reason to try to implement this feature. But I see it's pretty complicated and even if a good solution arises, it might not be worth it just because of the complexity.

Another idea is maybe that instead of updating all of the processes, only the new reservations added after the fact will pick up the changes & the current already scheduled ones will retain their original state? Or am I missing something why this would be an issue still? (Besides the fact that now we're having mixed configurations for different reservations/accounts)

**Comment by aaron-pham** [comment]
Btw, this is a really awesome project with very clean and readable code. I strive to write code as clean/nice as this!

**Comment by jdholtz** [comment]
> It's probably not a good idea to use multiprocessing.Manager (or other shared memory methods), huh.

Yeah, I would like to avoid using shared memory as much as possible. I'm not too familiar with it, as I have never dealt with shared memory, but it seems that the more processes using the same memory can have unintended consequences. Here is an excerpt from the [Python documentation](https://docs.python.org/2/library/multiprocessing.html#sharing-state-between-processes) on it:
```
When doing concurrent programming it is usually best to avoid using shared
state as far as possible. This is particularly true when using multiple processes.
```
If you do find a reliable (any relatively simple) way to do it though, you are welcome to create a PR.

> Any thoughts on this?

Hm, I don't see that being possible without saving the state in the script. Some possible solutions off the top of my head for your use case:
1. Filter the messages with your Discord bot
2. Run a new Docker container for every reservation/account added. This depends on how powerful the computer you are running the script on is (although the Docker container won't take up many resources) and how many flights you are running at once

> Another idea is maybe that instead of updating all of the processes, only the new reservations added after the fact will pick up the changes & the current already scheduled ones will retain their original state? Or am I missing something why this would be an issue still?

If implemented in this way, it would seem like a "half done feature" as it would partly work (register new reservations) and partly not (not updating the state of currently running accounts/reservations). 

> (Besides the fact that now we're having mixed configurations for different reservations/accounts)

I agree with this as a drawback as this would make it difficult for the user to understand what is going on (i.e. `check_fares` is set to `true`, but only for newly added reservations, whereas old reservations are still using `check_fares: false`).

> Btw, this is a really awesome project with very clean and readable code. I strive to write code as clean/nice as this!

I really appreciate that. Glad you are finding it useful.

**Comment by jdholtz** [comment]
@aaron-pham is this issue resolved?

**Comment by aaron-pham** [comment]
Yeah I guess we can close it. I may pursue some queue solution to feed the processes the updated data, but thinking about it off the top of my head makes it seem like somewhat of a complicated solution. Will message or create a PR if I find some simple solution.

**Comment by dtrachy** [comment]
You're right and I'm overthinking it. Besides, a systemd timer to bounce the script every day works just fine.
